src/utils/storageFacade.utils.ts

```diff
@@ -1,122 +1,355 @@
-    return data;
+    .eq('sessionid', sessionId)
+    .eq('package_ref', packageRef);
+
+  if (error) {
+    logger.error(error.message);
+    return null;
+  } else if (data) {
+    if (data.length == 0) {
+      //No entry found
+      logger.error('Storage Facade: Package "' + packageRef + '" not found');
+      return null;
+    } else if (data.length > 1) {
+      //Duplicates (how should we handle?)
+      logger.error(
+        'Storage Facade: Duplicate packages found for "' +
+          packageRef +
+          '" in session ' +
+          sessionId
+      );
+      return dbPkgToPkg(data[0]);
+    } else {
+      return dbPkgToPkg(data[0]);
+    }
+  }
+}
+
+/**
+ * Write Package, updates if entry exists in database, inserts otherwise
+ * @param pkg to write to database
+ * @param sessionId associated with one user
+ * @returns supabase write status
+ */
+export async function writePackage(pkg: Package, sessionId: number) {
+  let { data, error, status } = await supabase
+    .from('packages')
+    .select('*')
+    .eq('sessionid', sessionId)
+    .eq('package_ref', pkg.ref);
+
+  if (error) {
+    logger.error(error.message);
+  } else if (data) {
+    if (data.length == 0) {
+      //INSERT
+      return insertPackage(pkg, sessionId);
+    } else {
+      //UPDATE
+      return updatePackage(pkg, sessionId);
+    }
+  }
+  return status;
+}
+
+/**
+ * Read all vulnerabilities in one session
+ * @param sessionId associated with one user
+ * @returns list of vulnerabilities, empty list if none matching present in DB
+ */
+export async function readVulnsBySession(sessionId: number) {
+  let { data, error } = await supabase
+    .from('vulnerabilities')
+    .select('*,junction!inner(packageid,packages!inner(package_ref))')
+    .eq('junction.packages.sessionid', sessionId);
+
+  let cves: Vulnerability[] = [];
+  if (error) {
+    logger.error(error.message);
+  } else if (data) {
+    for (let v of data) {
+      cves.push(
+        //map database result to Vulnerability object
+        {
+          cveId: v.cveidstring,
+          packageRef: v.junction[0].packages.package_ref,
+          impact: v.impact,
+          likelihood: v.likelihood,
+          risk: v.risk,
+          cvss2: v.cvss_vector,
+        }
+      );
+    }
   }
-//Task 7: Double-check if read function overwrites current entry vs using update (applicable for packages & vulnerabilities)
-  //Task #5: Ensure sessionID is not one of the function parameters - need more clarity on sessionid usage before implementation
-/****************** WRITE PACKAGE**********************/
+
+  return cves;
+}
+
+/**
+ * Read all vulnerabilites in one package
+ * @param packageRef unique string ID retrieved from user-uploaded software bill of materials
+ * @param sessionId associated with one user
+ * @returns list of packages, empty list if none matching are present in DB
+ */
+export async function readVulnsByPkg(packageRef: string, sessionId: number) {
```

Comment:
The issue with left-joining vulnerabilities instead of packages is that it requires that the junction table have a session. Sessionid is only in packages, hence we need to reference first like so:

export async function readVulnBySession(sessionid:number){
    let {data} = await supabase
    .from ('packages')
    .select('junction!inner(packageid,vulnerabilities!inner(*))')
    .eq('sessionid', sessionid)
    return data;
    .eq('sessionid', sessionId)
    .eq('package_ref', packageRef);

And mind you, if we have multiple users who have the same cve for the same id, the junction table is supposed to have one entry for that. This methodology makes it so we'd have two sessioid columns that are not connected via foreign key so it'll cause more issues

Reply:
this is a nested junction, it joins vulnerabilities, junction, and packages
to ensure that you can read/write by package_ref 
You can test it in postman - I have a collection set up